src/wp-includes/webfonts-api/providers/class-wp-webfonts-local-provider.php

```diff
@@ -0,0 +1,98 @@
+<?php
+/**
+ * Webfonts API provider for locally-hosted fonts.
+ *
+ * @package    WordPress
+ * @subpackage WebFonts
+ * @since      5.9.0
+ */
+
+/**
+ * Webfonts API provider for locally-hosted fonts.
+ */
+final class WP_Webfonts_Local_Provider extends WP_Webfonts_Provider {
+
+	/**
+	 * The provider's unique ID.
+	 *
+	 * @since 5.9.0
+	 * @var string
+	 */
+	protected $id = 'local';
+
+	/**
+	 * Get validated params.
+	 *
+	 * @access public
+	 * @since 5.9.0
+	 * @param array $params The webfont's parameters.
+	 * @return array
+	 */
+	public function get_validated_params( $params ) {
+		$params = parent::get_validated_params( $params );
+
+		// Wrap font-family in quotes if it contains spaces.
+		if ( false !== strpos( $params['font-family'], ' ' ) && false === strpos( $params['font-family'], '"' ) && false === strpos( $params['font-family'], "'" ) ) {
+			$params['font-family'] = '"' . $params['font-family'] . '"';
+		}
+		return $params;
+	}
+
+	/**
+	 * Get the CSS for a collection of fonts.
+	 *
+	 * @access public
+	 * @since 5.9.0
+	 * @return string
+	 */
+	public function get_css() {
+		$css = '';
+		foreach ( $this->params as $font ) {
+
+			// Validate font params.
+			$font = $this->get_validated_params( $font );
+
+			if ( empty( $font['font-family'] ) ) {
+				continue;
+			}
+
+			$css .= '@font-face{';
+			foreach ( $font as $key => $value ) {
+
+				// Compile the "src" parameter.
+				if ( 'src' === $key ) {
+					$src = "local({$font['font-family']})";
+					foreach ( $value as $item ) {
+
+						// If the URL starts with "file:./" then it originated in a theme.json file.
```

Comment:
Hey, was looking at https://github.com/WordPress/gutenberg/issues/35591 and visited here out of curiosity. I don't know that anybody has started to look into `theme.json` integration. However, would it make sense to make this implementation "input agnostic"? It should no matter whether the theme registers the web fonts via `theme.json` or via the PHP functions. If the registration requires a path to the file, every consumer of this API should make sure the path is correct according to what this API asks (e.g.: the path is relative to the root theme directory, etc.).

Reply:
>"input agnostic"
>every consumer of this API should make sure the path is correct according to what this API asks (e.g.: the path is relative to the root theme directory, etc.).

Relative path to the theme makes sense for registering from the theme. What about a case where a font is also registered from a plugin or potentially a site has their assets outside of the theme and plugins?

Some ways to handle where the font file's root is located are:
- Assume the font file's source is from the theme's root unless an absolute path is given
- Add another parameter to the webfont's schema that can directly identify its root

Both of the above options eliminates the need to specify theme root location whether in the `theme.json` file or when using the API's `wp_register_webfonts()` function. And it provides flexibility for non-`theme.json` sites.

Reply:
Plugins and "classic" themes will be using the PHP functions, so they can just add the correct URL to the file using functions like `get_theme_file_uri()` or `plugins_url()`. In `theme.json` however we can't use these functions, so `file:./` was chosen as a way to allow defining local files while keeping things consistent with the way we define files in `block.json` files.

Reply:
Oh! OK... I understand what you mean now. Any ideas on how that would look in `theme.json`? Is it what @hellofromtonya mentioned above? 🤔 

> Add another parameter to the webfont's schema that can directly identify its root

Reply:
Thinking about this some more...
Perhaps that part of the code should not be in the webfonts API itself, but in the `theme.json` parser. The only reason I added it here was because the parser for `theme.json` currently doesn't support webfonts (WIP - discussion on https://github.com/WordPress/gutenberg/issues/35591). I needed to experiment and test things with the TT2 theme and see how we can do things with a structure that would be compatible with JSON values, so that piece of code was added.

If we move that code to the theme.json parser, then it can be removed from here and theme.json will properly take into account paths 🤔

Reply:
I see 2 options here:

1. Default to the theme's root as a base if the URL provided is not a full URL
2. Don't do anything for relative URLs in the API. The `file:./` implementation is only for the `theme.json` file, so it should live in the `theme.json` parser. The parser should convert relative paths to absolute URLs and pass them on to the webfonts API - which will only accept a full URL.

Reply:
After a discussion with @hellofromtonya yesterday about this, we'll be removing the `file:./` implementation from the API.
When we build the `theme.json` implementation, anything related to that will be in there.